src/hierarchies/Readme.md

```diff
@@ -0,0 +1,93 @@
+
+The class `AbstractHierarchy` defines the API, i.e. all the methods that need to be called 
+from outside of a Hierarchy class. 
+A template class `BaseHierarchy` inherits from `AbstractHierarchy` and implements some of the virtual methods in in, specifically: `sample_prior`, `sample_full_cond`, `initialize`, `add_datum`, `remove_datum`, `prior_pred_lpdf`, `conditional_pred_ldpf` `get_mutable_prior`, `like_lpdf_grid`, `prior_pred_lpdf_grid` and `conditional_pred_lpdf_grid`.
+These methods do not need to be implemented by the child classes. 
+
+Instead, child classes must implement:
+
+1. `like_lpdf`: evaluates k(x | theta_h)
+2. `marg_lpdf`: evaluates m(x) given some parameters theta_h (could be both the hyperparameters in P_0 or the paramters given by the full conditionals)
+3. `draw`: samples from P_0 given the parameters
+4. `clear_data`: clears all the summary statistics
+5. `update_hypers`: performs the update of parameters in P_0 given all the theta_h's (passed as a vector of protobuf Messages)
+6. `initialize_state`: initializes the current theta_h given the hyperparameters in P_0
+7. `initialize_hypers`: initializes the hyperparameters in P_0 given their hyperprior
+8. `update_summary_statistics`: updates the summary statistics when an observation is allocated or de-allocated from the hierarchy
+9. `get_posterior_parameters`: returns the paramters of the full conditional distribution **possible only when P_0 and k are conjugate**
+10. `set_state_from_proto`
+11. `write_state_to_proto`
+12. `write_hypers_to_proto`
+
+Observe that not all of these members are declared virtual in AbstractHierarchy or BaseHierarchy: this is because virtual members are only the ones that must be called from outside the Hierarchy, the other ones are handled via CRTP. Not having them virtual saves a lot of lookups in the vtables.
+
+The BaseHierarchy class takes 4 template parameters:
+1. `Derived` must be the type of the child class (needed for the CRTP)
+2. `State` is usually a struct representing theta_h
+3. `Hyperparams` is usually a struct representing the parameters in P_0
+4. `Prior` must be a protobuf object encoding the prior parameters.
+
+### Limitations
+
+Right now, we don't support nonconjugate hierarchies. In fact, the method `sample_full_cond` in `BaseHierarchy` does:
+1. performs the update of the parameters in the full conditional
+2. calls the method `draw` with the correct values of the parameters.
+
+Of course, this is valid only in P_0 and k are conjugate, since the posterior distribution will belong to the same family of P_0. In nonconjugate hierarchies a different update must be performed.
+
+A possibility is to derive a `ConjugateHierarchy` class from the `BaseHierarchy` one, that overrides `sample_full_cond` (performing what it does now in `BaseHierarchy`) and leave the possibility to override `sample_full_cond` with specific updates in the nonconjugate hierarchies.
```

Comment:
I think you mean `NonConjugateHierarchy`. Also, we could already implement such class if it isn't too much trouble, so that the user has no additional burden. Maybe even implemeting a non-conjugate hierarchy.

Reply:
No I meant that in `BaseHierarchy` we don't implement `sample_full_cond` thus leave it virtual as defined in `AbstractHierarchy`, while in `ConjugateHierarchy` we do the implementation.

Reply:
Right. Regardless, we could implement a `NonConjugateHierarchy` in the future

Reply:
Yes. I think it would be cool to have a Normal hierarchy with independent priors on the mean and the variance, so that we don't have that weird `var_scaling` parameter which is kind of a mess

Reply:
That is exactly the one I was thinking. I'll go ahead and create an issue for it, then mark this as resolved for now.